### ISSUE TITLE
ParagraphVector: Large Model Size when trying to save model to disk

### ISSUE THREAD

**Issue by grayhaired** [deeplearning4j/deeplearning4j#5691]
When I'm trying to save the paragraph model, it's taking too much of disk space. And as the model is loaded into memory when using, it's taking too much of memory. 
To be more specific, I've a text file of around 17 million keywords and I've created a paragraph vector model based on text file of size 324 MB.  And the paragraph vector model for this around 29 GB. Because of this, I've run this on a server of bigger RAM. This 29 GB is when I try to write with native java serializer. With dl4j serializer, it will come down to around 19 GB. 

Another problem is that even after long time, say one hour, it may not load the saved paragraphvector model. 

I face no issue if I train the model and run it to infer paragraph vector, without trying save the model to disk.  Only problem is when I try to save it or read a saved model. 

Now coming to the specifics of data and code - 

I can not paste sample text file here, but these are the statistics of that file - 

ls -l
-rw-r--r-- 1 my my 324M Jun 16 15:45 all.csv

wc -l all.csv 
17172944 all.csv

wc -w all.csv 
53472838 all.csv

Total Word Count 53477204; Line Count :17172944; Avg Word Count :3.1; Avg line Length 18.7 

Code to build Paragraph Vector model - 
` File file = new File("/home/my/all.csv");
        SentenceIterator iter = new BasicLineIterator(file);

        AbstractCache<VocabWord> cache = new AbstractCache<>();

        TokenizerFactory t = new DefaultTokenizerFactory();
        t.setTokenPreProcessor(new CommonPreprocessor());

        ParagraphVectors vec = new ParagraphVectors.Builder()
            .minWordFrequency(1)
            .iterations(5)
            .epochs(1)
            .layerSize(100)
            .learningRate(0.025)
            .windowSize(5)
            .iterate(iter)
            .trainWordVectors(true)
            .vocabCache(cache)
            .tokenizerFactory(t)
            .sampling(0)
            .build();
        vec.fit();

        try {
            FileOutputStream fileOut =
                new FileOutputStream("/home/my/para.pv");
            ObjectOutputStream out = new ObjectOutputStream(fileOut);
            out.writeObject(vec);
            out.close();
            fileOut.close();
            System.out.printf("Serialized data is saved in /home/my/para.pv");
        } catch (IOException i) {
            i.printStackTrace();
        }

14:47:39,176 INFO  ~ Loaded [CpuBackend] backend
14:47:39,503 INFO  ~ Number of threads used for NativeOps: 16
14:47:39,639 INFO  ~ Number of threads used for BLAS: 16
14:47:39,647 INFO  ~ Backend used: [CPU]; OS: [Linux]
14:47:39,648 INFO  ~ Cores: [32]; Memory: [67.6GB];
14:47:39,648 INFO  ~ Blas vendor: [OPENBLAS]
14:47:39,825 INFO  ~ Starting vocabulary building...
14:47:41,450 INFO  ~ Sequences checked: [100000]; Current vocabulary size: [138906]; Sequences/sec: 61576.35; Words/sec: 184062.81;
14:47:43,127 INFO  ~ Sequences checked: [200000]; Current vocabulary size: [254132]; Sequences/sec: 59630.29; Words/sec: 178438.88;
...
...
14:53:24,106 INFO  ~ Sequences checked: [17100000]; Current vocabulary size: [17654793]; Sequences/sec: 68870.52; Words/sec: 205361.57;
14:57:33,468 INFO  ~ Sequences checked: [17172944], Current vocabulary size: [17728930]; Sequences/sec: [28928.07];
14:57:33,479 INFO  ~ Projected memory use for model: [13.21 GB]
14:57:42,936 INFO  ~ Initializing syn1...
14:57:47,060 INFO  ~ Building learning algorithms:
14:57:47,060 INFO  ~           building SequenceLearningAlgorithm: [PV-DBOW]
14:57:47,063 INFO  ~           building ElementsLearningAlgorithm: [SkipGram]
14:57:47,063 INFO  ~ Starting learning process...
14:57:51,285 INFO  ~ Epoch: [1]; Words vectorized so far: [300268];  Lines vectorized so far: [100000]; Seq/sec: [23736.06]; Words/sec: [71271.78]; learningRate: [0.024978041258911173]
14:57:54,178 INFO  ~ Epoch: [1]; Words vectorized so far: [599661];  Lines vectorized so far: [200000]; Seq/sec: [34554.25]; Words/sec: [84376.11]; learningRate: [0.024956146069262265]
...
...
15:25:49,176 INFO  ~ Epoch: [1]; Words vectorized so far: [255429247];  Lines vectorized so far: [85500000]; Seq/sec: [76452.60]; Words/sec: [151850.95]; learningRate: [0.006320027251208471]
15:25:50,420 INFO  ~ Epoch: [1]; Words vectorized so far: [255728296];  Lines vectorized so far: [85600000]; Seq/sec: [80385.85]; Words/sec: [151916.39]; learningRate: [0.006298158242705127]
15:25:51,523 INFO  ~ Epoch [1] finished; Elements processed so far: [255984340];  Sequences processed: [85685905]
15:25:51,523 INFO  ~ Time spent on training: 1684460 ms
...
...
15:25:49,176 INFO  ~ Epoch: [1]; Words vectorized so far: [255429247];  Lines vectorized so far: [85500000]; Seq/sec: [76452.60]; Words/sec: [151850.95]; learningRate: [0.006320027251208471]
15:25:50,420 INFO  ~ Epoch: [1]; Words vectorized so far: [255728296];  Lines vectorized so far: [85600000]; Seq/sec: [80385.85]; Words/sec: [151916.39]; learningRate: [0.006298158242705127]
15:25:51,523 INFO  ~ Epoch [1] finished; Elements processed so far: [255984340];  Sequences processed: [85685905]
15:25:51,523 INFO  ~ Time spent on training: 1684460 ms

I'm running on debian 9 machine.

**Comment by raver119** [comment]
So, what i did. I've wrote simple tests where rather big INDArray is created, then serialized and deserialized.

Here's output:
```
o.n.l.s.LargeSerDeTests - Shape: [100000, 12500]; Length: 1250000000
o.n.l.s.LargeSerDeTests - Starting serialization...
o.n.l.s.LargeSerDeTests - Starting deserialization...
o.n.l.a.b.BaseDataBuffer - Loading a data stream with opType different from what is set globally. Expect precision loss
o.n.l.s.LargeSerDeTests - Timings: {Ser : 85326 ms; De: 22033 ms;}
```

That basically converts to 220MB/s for deserialization. and 55MB/s for serialization. Both values look reasonable for me. Amount of data is reasonable as well: 5GB file representing model.

**Comment by raver119** [comment]
Now, please run the same test on the same box where you've got problems. And we'll see numbers 😄 

https://gist.github.com/raver119/5efefbe83e87ffc34ebcc031c4bdca64

**Comment by raver119** [comment]
P.s. `val` comes from lombok, you can just replace it with manual variable types declaration.

**Comment by raver119** [comment]
@grayhaired any news?

**Comment by grayhaired** [comment]
Sorry for the delay.  Here is the run. I've posted additional details so that you know about system I'm running. I'm actually running this on Ubuntu. ( I wrote it as debian by mistake)

o.n.l.f.Nd4jBackend - Loaded [CpuBackend] backend
o.n.n.NativeOpsHolder - Number of threads used for NativeOps: 16
o.n.n.Nd4jBlas - Number of threads used for BLAS: 16
o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CPU]; OS: [Linux]
o.n.l.a.o.e.DefaultOpExecutioner - Cores: [32]; Memory: [40.0GB];
o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [OPENBLAS]
Shape: 100000,12500; Length: 1250000000
Starting serialization...
Starting deserialization...
Timings: Ser : 146734, DeSer :34833
